sklearn/manifold/_locally_linear.py

```diff
@@ -415,9 +419,9 @@ def locally_linear_embedding(
         # find the eigenvectors and eigenvalues of each local covariance
         # matrix. We want V[i] to be a [n_neighbors x n_neighbors] matrix,
         # where the columns are eigenvectors
-        V = np.zeros((N, n_neighbors, n_neighbors))
+        V = np.zeros((N, n_neighbors, n_neighbors), dtype=X.dtype)
         nev = min(d_in, n_neighbors)
-        evals = np.zeros([N, nev])
+        evals = np.zeros([N, nev], dtype=X.dtype)
 
         # choose the most efficient way to find the eigenvectors
         use_svd = n_neighbors > d_in
```

Comment:
We should modify the creation of `tmp` to avoid some later casting:

```python
tmp = np.dot(V.transpose(0, 2, 1), np.ones(n_neighbors, dtype=X.dtype))

Reply:
The same with the initialization of `w_reg`:

```python
w_reg = np.zeros((N, n_neighbors), dtype=X.dtype)